test/common/upstream/cds_speed_test.cc

```diff
@@ -0,0 +1,190 @@
+
+    // make a pile of dynamic clusters and add them to the response
+    for (size_t i = 0; i < num_clusters; ++i) {
+      envoy::config::endpoint::v3::ClusterLoadAssignment cluster_load_assignment;
+      cluster_load_assignment.set_cluster_name("fare_" + std::to_string(i));
+
+      auto* endpoints = cluster_load_assignment.add_endpoints();
+      endpoints->set_priority(1);
+      auto* locality = endpoints->mutable_locality();
+      locality->set_region("region");
+      locality->set_zone("zone");
+      locality->set_sub_zone("sub_zone");
+
+      auto* resource = response->mutable_resources()->Add();
+      resource->PackFrom(cluster_load_assignment);
+      if (v2_config_) {
+        RELEASE_ASSERT(resource->type_url() ==
+                           "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
+                       "");
+        resource->set_type_url("type.googleapis.com/envoy.api.v2.ClusterLoadAssignment");
+      }
+    }
+    validation_visitor_.setSkipValidation(ignore_unknown_dynamic_fields);
+
+    state_.SetComplexityN(num_clusters);
+    state_.ResumeTiming();
+    grpc_mux_->grpcStreamForTest().onReceiveMessage(std::move(response));
+    state_.PauseTiming();
+  }
+
+  TestDeprecatedV2Api _deprecated_v2_api_;
+  State& state_;
+  const bool v2_config_;
+  const std::string type_url_;
+  uint64_t version_{};
+  bool initialized_{};
+  Stats::IsolatedStoreImpl stats_;
+  Config::SubscriptionStats subscription_stats_;
+  Ssl::MockContextManager ssl_context_manager_;
+  envoy::config::cluster::v3::Cluster eds_cluster_;
+  NiceMock<MockClusterManager> cm_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  EdsClusterImplSharedPtr cluster_;
+  Config::SubscriptionCallbacks* eds_callbacks_{};
+  Config::OpaqueResourceDecoderImpl<envoy::config::endpoint::v3::ClusterLoadAssignment>
+      resource_decoder_{validation_visitor_, "cluster_name"};
+  NiceMock<Random::MockRandomGenerator> random_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Server::MockAdmin> admin_;
+  Singleton::ManagerImpl singleton_manager_{Thread::threadFactoryForTest()};
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  ProtobufMessage::MockValidationVisitor validation_visitor_;
+  Api::ApiPtr api_;
+  Grpc::MockAsyncClient* async_client_;
+  NiceMock<Grpc::MockAsyncStream> async_stream_;
+  Config::GrpcMuxImplSharedPtr grpc_mux_;
+  Config::GrpcSubscriptionImplPtr subscription_;
+};
+
+} // namespace Upstream
+} // namespace Envoy
+
+static void addClusters(State& state) {
+  Envoy::Upstream::CdsSpeedTest speed_test(state, state.range(0));
+  for (auto _ : state) { // NOLINT(clang-analyzer-deadcode.DeadStores)
+    // timing is resumed within the helper:
+    state.PauseTiming();
+    // if we've been instructed to skip tests, only run once no matter the argument:
+    uint32_t clusters = skipExpensiveBenchmarks() ? 1 : state.range(2);
```

Comment:
nit: name this "num_clusters" as if it's just "clusters" it sounds like a container. Also hoist this out of the loop, and make it const :)